lib/vector-config/tests/basic.rs

```diff
@@ -0,0 +1,235 @@
+        max_events: Some(10000),
+        max_bytes: Some(16_000_000),
+        timeout: Some(SpecialDuration(5)),
+    }
+}
+
+const fn default_simple_sink_encoding() -> Encoding {
+    Encoding::Json { pretty: true }
+}
+
+fn default_simple_sink_endpoint() -> String {
+    String::from("https://zalgo.io")
+}
+
+/// A sink for sending events to the `advanced` service.
+#[derive(Clone)]
+#[configurable_component(sink)]
+#[configurable(metadata(status = "stable"))]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub struct AdvancedSinkConfig {
+    /// The endpoint to send events to.
+    #[serde(default = "default_advanced_sink_endpoint")]
+    endpoint: String,
+    /// The agent version to simulate when sending events to the downstream service.
+    ///
+    /// Must match the pattern of "v\d+\.\d+\.\d+", which allows for values such as `v1.23.0` or `v0.1.3`, and so on.
+    #[configurable(validation(pattern = "foo"))]
+    agent_version: String,
+    #[configurable(derived)]
+    #[serde(default = "default_advanced_sink_batch")]
+    batch: BatchConfig,
+    #[configurable(derived)]
+    #[deprecated]
+    #[serde(default = "default_advanced_sink_encoding")]
+    encoding: Encoding,
+    /// The tags to apply to each event.
+    tags: HashMap<String, String>,
+}
+
+const fn default_advanced_sink_batch() -> BatchConfig {
+    BatchConfig {
+        max_events: Some(5678),
+        max_bytes: Some(36_000_000),
+        timeout: Some(SpecialDuration(15)),
+    }
+}
+
+const fn default_advanced_sink_encoding() -> Encoding {
+    Encoding::Json { pretty: true }
+}
+
+fn default_advanced_sink_endpoint() -> String {
+    String::from("https://zalgohtml5.io")
+}
+
+/// Collection of various sources available in Vector.
+#[derive(Clone)]
+#[configurable_component]
+pub enum SourceConfig {
+    /// Simple source.
+    Simple(#[configurable(derived)] SimpleSourceConfig),
+}
+
+/// Collection of various sinks available in Vector.
+#[derive(Clone)]
+#[configurable_component]
+pub enum SinkConfig {
+    /// Simple sink.
+    Simple(#[configurable(derived)] SimpleSinkConfig),
+
+    /// Advanced sink.
+    Advanced(#[configurable(derived)] AdvancedSinkConfig),
+}
+
+#[derive(Clone)]
+#[configurable_component]
+#[configurable(description = "Global options for configuring Vector.")]
+pub struct GlobalOptions {
+    /// The data directory where Vector will store state.
+    data_dir: Option<String>,
+}
+
+/// The overall configuration for Vector.
+#[derive(Clone)]
+#[configurable_component]
+pub struct VectorConfig {
+    #[configurable(derived)]
+    global: GlobalOptions,
+    /// Any configured sources.
+    sources: Vec<SourceConfig>,
+    /// Any configured sinks.
+    sinks: Vec<SinkConfig>,
+}
+
+#[test]
+fn vector_config() {
+    let root_schema = generate_root_schema::<VectorConfig>();
+    let json = serde_json::to_string_pretty(&root_schema)
+        .expect("rendering root schema to JSON should not fail");
+
+    println!("{}", json);
```

Comment:
Do we want to assert against a fixed schema or something like that?

Reply:
Ah, right, so this test was basically just a harness for me to easily test changes to the macro as I was developing. As part of the larger integration PR that will follow, I do plan to have actual unit tests to assert the generated schema.

I mostly figured that likelihood of things changing when actually integrating it would be... very high, and so having to tweak all of those unit tests would just be laborious, since I wasn't sure what we need to be added, changed, or removed after optimizing towards how it actually gets used.